src/client/MainnetDataDownloader.ts

```diff
@@ -0,0 +1,553 @@
+
+      if (toBlock == "afterInitialization") return;
+
+      // download events after initialization
+      await this.downloadEvents(
+        uniswapV3Pool,
+        eventDB,
+        initializationEventBlockNumber,
+        toBlockAsNumber,
+        batchSize
+      );
+
+      await this.preProcessSwapEvent(eventDB);
+    } finally {
+      await eventDB.close();
+    }
+  }
+
+  async update(
+    mainnetEventDBFilePath: string,
+    toBlock: EndBlockTypeWhenRecover,
+    batchSize: number = 500
+  ) {
+    // check dbfile first
+    let { poolAddress } = this.parseFromMainnetEventDBFilePath(
+      mainnetEventDBFilePath
+    );
+    let dbExists = false;
+    try {
+      accessSync(mainnetEventDBFilePath, constants.F_OK);
+      dbExists = true;
+    } catch (err) {}
+    if (!dbExists)
+      throw new Error("DB file doesn't exist. Please download it.");
+
+    // check toBlock then
+    let deploymentBlockNumber = await this.queryDeploymentBlockNumber(
+      poolAddress
+    );
+    let toBlockAsNumber = await this.parseEndBlockTypeWhenRecover(
+      toBlock,
+      poolAddress
+    );
+    if (toBlockAsNumber < deploymentBlockNumber)
+      throw new Error("toBlock is too small, the pool hasn't been deployed.");
+
+    let eventDB = await EventDBManager.buildInstance(mainnetEventDBFilePath);
+    try {
+      let latestEventBlockNumber = await eventDB.getLatestEventBlockNumber();
+
+      if (toBlockAsNumber < latestEventBlockNumber) {
+        return;
+      }
+
+      let uniswapV3Pool = await this.getCorePoolContarct(poolAddress);
+
+      // check and record initialize event if needed
+      let updateInitializationEvent = false;
+      let initializationEventBlockNumber =
+        await eventDB.getInitializationEventBlockNumber();
+      if (0 == initializationEventBlockNumber) {
+        updateInitializationEvent = true;
+        let initializeTopic = uniswapV3Pool.filters.Initialize();
+        let initializationEvent = await uniswapV3Pool.queryFilter(
+          initializeTopic
+        );
+        await eventDB.addInitialSqrtPriceX96(
+          initializationEvent[0].args.sqrtPriceX96.toString()
+        );
+        initializationEventBlockNumber = initializationEvent[0].blockNumber;
+        await eventDB.saveInitializationEventBlockNumber(
+          initializationEventBlockNumber
+        );
+        await eventDB.saveLatestEventBlockNumber(
+          initializationEventBlockNumber
+        );
+      }
+
+      // download events after initialization
+      await this.downloadEvents(
+        uniswapV3Pool,
+        eventDB,
+        updateInitializationEvent
+          ? initializationEventBlockNumber
+          : latestEventBlockNumber + 1,
+        toBlockAsNumber,
+        batchSize
+      );
+
+      await this.preProcessSwapEvent(eventDB);
+    } finally {
+      await eventDB.close();
+    }
+  }
+
+  async initializeAndReplayEvents(
```

Comment:
Should we call this `PreProcessAndReplayEvents` instead?

Reply:
Are you going to say "or" instead of "and"? It's for both preProcess and replayEvents use and I'm not sure if that's ok. If not, I think we can do some refactor to separate them. And we'd better add not only amountSpecified but sqrtPriceLimitX96 in events DB as intact input to replay a swap event.